File: services/auth-service/src/domain/use-cases/LoginUseCase.ts
import { generarToken, verificarPassword, UnauthorizedError } from '@sec/shared';
import { IUsuarioRepository } from '../repositories/IUsuarioRepository';

interface LoginInput {
  email:    string;
  password: string;
}

interface LoginOutput {
  token: string;
  user: {
    id:          number;
    nombre:      string;
    email:       string;
    rol:         string;
    sedeId:      number;
    sedeCodigo:  string;
    zona:        string | null;
  };
}

/**
 * Caso de uso: Login
 * Valida credenciales y genera el JWT con sedeId para multi-tenancy.
 * No conoce Express ni Oracle — solo habla con el repositorio (interfaz).
 */
export class LoginUseCase {
  constructor(private readonly usuarioRepo: IUsuarioRepository) {}

  async execute(input: LoginInput): Promise<LoginOutput> {
    const { email, password } = input;

    const usuario = await this.usuarioRepo.findByEmail(email);

    if (!usuario) {
      throw new UnauthorizedError('Credenciales inválidas');
    }

    if (!usuario.estaActivo()) {
      throw new UnauthorizedError('Usuario inactivo. Contacte al administrador.');
    }

    const passwordValido = await verificarPassword(password, usuario.passwordHash);
    if (!passwordValido) {
      throw new UnauthorizedError('Credenciales inválidas');
    }

    const token = generarToken({
      userId:     usuario.id,
      email:      usuario.email,
      rol:        usuario.rol,
      sedeId:     usuario.sedeId,
      sedeCodigo: usuario.sedeCodigo,
      regionId:   usuario.regionId,
    });

    return {
      token,
      user: {
        id:         usuario.id,
        nombre:     usuario.nombreCompleto,
        email:      usuario.email,
        rol:        usuario.rol,
        sedeId:     usuario.sedeId,
        sedeCodigo: usuario.sedeCodigo,
        zona:       usuario.zonaAsignada,
      },
    };
  }
}

File: services/auth-service-py/src/domain/use_cases/login_use_case.py
from dataclasses import dataclass
from typing import Optional
import bcrypt

from ..repositories.i_usuario_repository import IUsuarioRepository


@dataclass
class LoginInput:
    email: str
    password: str


@dataclass
class LoginOutput:
    token: str
    user: dict


class UnauthorizedError(Exception):
    pass


class LoginUseCase:
    def __init__(self, usuario_repo: IUsuarioRepository, generar_token_fn):
        self._repo = usuario_repo
        self._generar_token = generar_token_fn

    async def execute(self, input: LoginInput) -> LoginOutput:
        usuario = await self._repo.find_by_email(input.email)

        if not usuario:
            raise UnauthorizedError("Credenciales inválidas")

        if not usuario.esta_activo():
            raise UnauthorizedError("Usuario inactivo. Contacte al administrador.")

        if not bcrypt.checkpw(input.password.encode(), usuario.password_hash.encode()):
            raise UnauthorizedError("Credenciales inválidas")

        token = self._generar_token({
            "userId":     usuario.id,
            "email":      usuario.email,
            "rol":        usuario.rol,
            "sedeId":     usuario.sede_id,
            "sedeCodigo": usuario.sede_codigo,
            "regionId":   usuario.region_id,
        })

        return LoginOutput(
            token=token,
            user={
                "id":         usuario.id,
                "nombre":     usuario.nombre_completo,
                "email":      usuario.email,
                "rol":        usuario.rol,
                "sedeId":     usuario.sede_id,
                "sedeCodigo": usuario.sede_codigo,
                "zona":       usuario.zona_asignada,
            }
        )
